Makefile

```diff
@@ -1,58 +1,313 @@
+
+      ifeq (,$(_UNAME_ARCH))
+	$(warning WARNING: Can not detect system architecture!)
+      endif
+    endif
+
+    ifeq ($(_UNAME_ARCH),x86_64)
+	CANDIDATE_LIBDIR = $(PREFIX)/lib64
+    endif
+  endif
+endif
 
-INSTALL ?= install
+LIBDIR ?= $(CANDIDATE_LIBDIR)
 
-ifeq ($(shell uname -s),Darwin)
-	SHARED_NAME := libkissfft.dylib
-	SHARED_FLAGS := -Wl,-install_name,$(SHARED_NAME)
+export ABS_LIBDIR = $(abspath $(LIBDIR))
+
+export INSTALL ?= install
+
+#
+# Library name and version
+#
+
+ifeq ($(KISSFFT_OPENMP), 1)
+	KISSFFTLIB_SHORTNAME = kissfft-$(KISSFFT_DATATYPE)-openmp
+	KISSFFT_PKGCONFIG = kissfft-$(KISSFFT_DATATYPE)-openmp.pc
+	KISSFFTLIB_FLAGS = -fopenmp
+	TYPEFLAGS = -fopenmp
+	PKGCONFIG_OPENMP = -fopenmp
 else
-	SHARED_NAME := libkissfft.so
-	SHARED_FLAGS := -Wl,-soname,$(SHARED_NAME)
+	KISSFFTLIB_SHORTNAME = kissfft-$(KISSFFT_DATATYPE)
+	KISSFFT_PKGCONFIG = kissfft-$(KISSFFT_DATATYPE).pc
+	TYPEFLAGS =
+	PKGCONFIG_OPENMP =
 endif
 
-all:
-	$(CC) -Wall -fPIC -c *.c -Dkiss_fft_scalar=$(DATATYPE) -o kiss_fft.o
-	$(AR) crus libkissfft.a kiss_fft.o
-	$(CC) -shared $(SHARED_FLAGS) -o $(SHARED_NAME) kiss_fft.o
+ifeq ($(KISSFFT_STATIC), 1)
+	KISSFFTLIB_NAME = lib$(KISSFFTLIB_SHORTNAME).a
+	KISSFFTLIB_FLAGS += -static
+else ifeq ($(shell uname -s),Darwin)
+	KISSFFTLIB_NAME = lib$(KISSFFTLIB_SHORTNAME).dylib
+	KISSFFTLIB_FLAGS += -shared -Wl,-install_name,$(KISSFFTLIB_NAME)
+else
+	KISSFFTLIB_SODEVELNAME = lib$(KISSFFTLIB_SHORTNAME).so
+	KISSFFTLIB_SONAME = $(KISSFFTLIB_SODEVELNAME).$(KFVER_MAJOR)
+	KISSFFTLIB_NAME = $(KISSFFTLIB_SONAME).$(KFVER_MINOR).$(KFVER_PATCH)
+	KISSFFTLIB_FLAGS += -shared -Wl,-soname,$(KISSFFTLIB_SONAME)
+endif
+
+export KISSFFTLIB_SHORTNAME
+
+#
+# Compile-time definitions by datatype
+#
+#
+# Note: -DKISS_FFT_BUILD and -DKISS_FFT_SHARED control
+# C symbol visibility.
+#
+
+ifeq "$(KISSFFT_DATATYPE)" "int32_t"
+	TYPEFLAGS += -DFIXED_POINT=32
+else ifeq "$(KISSFFT_DATATYPE)" "int16_t"
+	TYPEFLAGS += -DFIXED_POINT=16
+else ifeq "$(KISSFFT_DATATYPE)" "simd"
+	TYPEFLAGS += -DUSE_SIMD=1 -msse
+else ifeq "$(KISSFFT_DATATYPE)" "float"
+	TYPEFLAGS += -Dkiss_fft_scalar=$(KISSFFT_DATATYPE)
+else ifeq "$(KISSFFT_DATATYPE)" "double"
+	TYPEFLAGS += -Dkiss_fft_scalar=$(KISSFFT_DATATYPE)
+else
+	$(error ERROR: KISSFFT_DATATYPE must be one of: float double int16_t int32_t simd)
+endif
+
+ifneq ($(KISSFFT_STATIC), 1)
+	TYPEFLAGS += -DKISS_FFT_SHARED
+endif
+
+ifeq ($(KISSFFT_USE_ALLOCA), 1)
+	TYPEFLAGS += -DKISS_FFT_USE_ALLOCA=1
+endif
+
+#
+# Compile-time definitions
+#
+
+#
+# Save pkgconfig variables before appending
+# -DKISS_FFT_BUILD to TYPEFLAGS
+#
+
+ifneq ($(shell uname -s),Darwin)
+	PKGCONFIG_KISSFFT_VERSION = $(KFVER_MAJOR).$(KFVER_MINOR).$(KFVER_PATCH)
+	PKGCONFIG_KISSFFT_OUTPUT_NAME = $(KISSFFTLIB_SHORTNAME)
+	PKGCONFIG_PKG_KISSFFT_DEFS = $(TYPEFLAGS)
+	PKGCONFIG_KISSFFT_PREFIX = $(ABS_PREFIX)
+	PKGCONFIG_KISSFFT_INCLUDEDIR = $${prefix}/include
+	PKGCONFIG_KISSFFT_LIBDIR = $${prefix}/lib
```

Comment:
Hmm, this will be problematic when `CANDIDATE_LIBDIR` matches `lib64`. Which can be fixed by using `CANDIDATE_LIBDIR_NAME=lib64` and and `PKGCONFIG_KISSFFT_LIBDIR = $${prefix}/$(CANDIDATE_LIBDIR_NAME)`.

But it gets even worse when `LIBDIR` is overridden (e.g. debian uses [`--libdir=\$${prefix}/lib/$(ARCH)`](https://wiki.debian.org/Multiarch/Implementation)). In the absence of `libdir` variable relative to `prefix` like CMake’s `GNUInstallDirs` module provides us, maybe something like

```make
ifeq "$(LIBDIR)" "$(PREFIX)/$(CANDIDATE_LIBDIR_NAME)"
 	PKGCONFIG_KISSFFT_PKGINCLUDEDIR = $${prefix}/$(CANDIDATE_LIBDIR_NAME)
else
 	PKGCONFIG_KISSFFT_PKGINCLUDEDIR = $(LIBDIR)
endif
```

would be good enough substitute.

Reply:
Should I also substritute `$(INCDIR)` too for the same reason? I guess so. $(`BINDIR)` does not land in pkg-config file so it should be left intaxt.

Reply:
Yes, the `includedir` suffers from the same issue (sans candidate selection). We actually change that one on Nix to a different prefix.

Reply:
Uploaded the fix + fixed `$(FFTWLIB)` in `test/Makefile`. Hopefully this time it is OK :)